File: frontend/src/app/core/services/audit.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface AuditLog {
  id: number;
  user_id: number;
  contract_id: number | null;
  entity_type: string | null;
  entity_id: number | null;
  action: string;
  details: string | null;
  created_at: string;
}

export interface AuditResponse {
  value: AuditLog[];
  Count: number;
}

@Injectable({
  providedIn: 'root'
})
export class AuditService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = environment.apiUrl;

  getAuditLogs(): Observable<AuditResponse> {
    return this.http.get<AuditResponse>(
      `${this.apiUrl}/audit-logs`
    );
  }
}

File: app/services/audit_service.py
from sqlalchemy.orm import Session

from app.models.audit_log import AuditLog


def create_audit_log(
    db: Session,
    user_id: int,
    action: str,
    entity_type: str,
    entity_id: int,
    contract_id: int | None = None,
    details: str | None = None,
):
    audit_log = AuditLog(
        user_id=user_id,
        contract_id=contract_id,
        action=action,
        entity_type=entity_type,
        entity_id=entity_id,
        details=details,
    )

    db.add(audit_log)

    return audit_log